ocaml/tests/test_clustering.ml

```diff
@@ -376,12 +376,205 @@ let test_disallow_unplug_ro_with_clustering_enabled =
   ; "test_disallow_unplug_with_clustering", `Quick, test_disallow_unplug_with_clustering
   ]
 
+let default = !Xapi_globs.cluster_stack_default
+let default_smapiv3 = Constants.default_smapiv3_cluster_stack
+let test_default = "default_sm_stack_value_used_in_place_of_xhad"
+
+let choose_cluster_stack_should_select cluster_stack ~__context =
+  Alcotest.(check string) "choose_cluster_stack"
+  cluster_stack
+  (Cluster_stack_constraints.choose_cluster_stack ~__context)
+
+let choose_cluster_stack_should_fail_with_conflict ~__context =
+  Alcotest.check_raises
+    "choose_cluster_stack should fail with different cluster_stacks provided"
+    (Failure "Conflicting cluster stack demands.")
+    (fun () -> Cluster_stack_constraints.choose_cluster_stack ~__context |> ignore)
+
+(* Choose_cluster_stack looks at SM types and their required_cluster_stacks
+ * If any SRs of matching type are found, the first corresponding cluster_stack is returned,
+ * If no SRs are found it returns the existing cluster's stack, otherwise it returns the
+ * default cluster_stack in Xapi_globs *)
+
+let test_choose_cluster_stack_clusters_no_sms () =
+  let __context = T.make_test_database () in
+  choose_cluster_stack_should_select default ~__context;
+
+  (* Add two cluster, test choose_cluster_stack's filtering *)
+  for i = 0 to 1 do
+    let _ = T.make_cluster_and_cluster_host ~__context () in
+    choose_cluster_stack_should_select default_smapiv3 ~__context
+  done;
+
+  (* choose_cluster_stack should fail with multiple different cluster_stacks *)
+  let _ = T.make_cluster_and_cluster_host ~__context ~cluster_stack:"different_stack" () in
```

Comment:
I think the creation would fail in a real situation here because you can't have more than one cluster stack, so this last part of test should probably be removed.